File: lib/other/src/main/java/ch/unstable/ost/api/search/types/ConnectionDeserializer.java
package ch.unstable.ost.api.search.types;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

import ch.unstable.ost.api.model.Connection;
import ch.unstable.ost.api.model.Section;

public class ConnectionDeserializer implements JsonDeserializer<Connection> {
    @Override
    public Connection deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonObject connectionObj = json.getAsJsonObject();
        Section[] sections = context.deserialize(connectionObj.get("legs"), Section[].class);
        return new Connection(sections);
    }
}

File: app/src/main/java/ch/unstable/ost/api/search/types/ConnectionDeserializer.kt
package ch.unstable.ost.api.search.types

import ch.unstable.ost.api.base.SectionListDeserializer
import ch.unstable.ost.api.model.Connection
import ch.unstable.ost.api.model.Section
import com.google.gson.*
import java.lang.reflect.Type

object ConnectionDeserializer : JsonDeserializer<Connection> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Connection {
        val connectionObj :JsonObject = json.asJsonObject
        val sections: List<Section> = context.deserialize(connectionObj.get("legs"), SectionListDeserializer.type)
        return Connection(sections)
    }
}